### ISSUE TITLE
Project: Add Expressions to JX Library

### ISSUE THREAD

**Issue by dthain** [cooperative-computing-lab/cctools#1073]
Extend the JX library to permit simple expressions that can be evaluated within the context of another JSON document.  This will make it handy to filter and evaluate JSON data wherever we use it, so that things like this become easy:

```
chirp_status --where '(avail/total)>.9`
```

Here is the idea: The function `jx_evaluate` should take an arbitrary JSON expression, possibly containing symbols and operators, look them up in the JSON object `context`, and then return a new expression with everything resolved, like this:

```
struct jx * jx_evaluate( struct jx * expr, struct jx *context );
```

For example, suppose that `context` is this:

```
`{ "name":"honda", "mass": 1800, "speed": 65.7 }`
```

Then `name` would evaluate to `"honda"`, `speed>10` would evaluate to `true`, and `{ "energy": mass*speed*speed/2.0}` would evaluate to `{"energy":3884760}`

Here is a plan to go about it:

1 - Add new `JX_EXPR` node type to `jx.[ch]` and enumerate the various operators, so that you can create expressions manually like this:

```
j = jx_expr( jx_integer(10), JX_MULTIPLY, jx_symbol(speed) );
```

2 - Modify `jx_print.[ch]` to accommodate printing of expressions.
3 - Modify `jx_parse.c` to parse expressions using a simple recursive descent parser.
4 - Modify `jx_evaluate.c` to evaluate expressions down to constant values.

Once that is all working, then you can easily modify `chirp_status` to take an optional command-line argument, and display only the objects that match.

**Comment by dthain** [comment]
And the follow up to this project is that the same expressions can then be used to generate arbitrary queries in DeltaDB.

**Comment by dthain** [comment]
Almost done here.  They last major piece missing are we need to be able to evaluate some functions, particularly:

- `range(a,b)` returns `[a, a+1, ... b]`
- `regex(string,regex)` returns true if string matches regex
- `foreach(x,array,expr)` evaluates expr for each value of x in array.

Once we can do that, then we can represent a range of jobs with syntax like this:

```
foreach(x,range(1,10),
{
   cmd: "sim.exe -p "+(x*2)+" -o output."+x,
   output: "output."+x
   . . .
}
)
```

**Comment by dthain** [comment]
@trshaffer What's the current status of JX expressions in Makeflow?

**Comment by trshaffer** [comment]
We have an experimental implementation that can do things, but I'm not really happy with the syntax. Since it's not an official feature, I guess now's the time for breaking changes. I was thinking of switching to something that looks more like a scripting language a la that OpenSCAD you mentioned, over the current Lisp-y form. Maybe we should have a meeting with @btovar, @nhazekam, etc. to discuss?

**Comment by dthain** [comment]
Ok, why don't you go ahead and sketch out your thougts here in the PR, and then we can discuss.

**Comment by dthain** [comment]
Implemented in PR #1683
